

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>tirank.TrainPre &mdash; TiRank 1.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=e59714d7" />

  
      <script src="../_static/jquery.js?v=5d32c60e"></script>
      <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../_static/documentation_options.js?v=f2a433a1"></script>
      <script src="../_static/doctools.js?v=9a2dae69"></script>
      <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="tirank.TrainPre.AssignPcluster" href="tirank.TrainPre.AssignPcluster.html" />
    <link rel="prev" title="tirank.Model.TransformerEncoderModel" href="tirank.Model.TransformerEncoderModel.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            TiRank
              <img src="../_static/TiRank_white.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Getting Started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../model_input.html">Model Input</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">CLI Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../tutorial_st_survival.html">Tutorial 1: Spatial Transcriptomics (ST) + Cox Survival Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial_sc_classification.html">Tutorial 2: scRNA-seq + Classification Analysis</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">GUI Tutorial</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../tutorial_web.html">TiRank Web App Tutorial</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Features Document</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../result_interpretation.html">Result Interpretation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../hyperparameters.html">Hyperparameters</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API Reference</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../api.html">API Reference</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../api.html#core-model-training">Core Model &amp; Training</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="tirank.Model.html">tirank.Model</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">tirank.TrainPre</a><ul>
<li class="toctree-l4"><a class="reference internal" href="tirank.TrainPre.AssignPcluster.html">tirank.TrainPre.AssignPcluster</a></li>
<li class="toctree-l4"><a class="reference internal" href="tirank.TrainPre.IdenHub.html">tirank.TrainPre.IdenHub</a></li>
<li class="toctree-l4"><a class="reference internal" href="tirank.TrainPre.Pcluster.html">tirank.TrainPre.Pcluster</a></li>
<li class="toctree-l4"><a class="reference internal" href="tirank.TrainPre.Predict.html">tirank.TrainPre.Predict</a></li>
<li class="toctree-l4"><a class="reference internal" href="tirank.TrainPre.Reject_With_GMM_Bio.html">tirank.TrainPre.Reject_With_GMM_Bio</a></li>
<li class="toctree-l4"><a class="reference internal" href="tirank.TrainPre.Reject_With_GMM_Reg.html">tirank.TrainPre.Reject_With_GMM_Reg</a></li>
<li class="toctree-l4"><a class="reference internal" href="tirank.TrainPre.Reject_With_StrictNumber.html">tirank.TrainPre.Reject_With_StrictNumber</a></li>
<li class="toctree-l4"><a class="reference internal" href="tirank.TrainPre.Train_one_epoch.html">tirank.TrainPre.Train_one_epoch</a></li>
<li class="toctree-l4"><a class="reference internal" href="tirank.TrainPre.Validate_model.html">tirank.TrainPre.Validate_model</a></li>
<li class="toctree-l4"><a class="reference internal" href="tirank.TrainPre.get_best_model.html">tirank.TrainPre.get_best_model</a></li>
<li class="toctree-l4"><a class="reference internal" href="tirank.TrainPre.objective.html">tirank.TrainPre.objective</a></li>
<li class="toctree-l4"><a class="reference internal" href="tirank.TrainPre.permute_once.html">tirank.TrainPre.permute_once</a></li>
<li class="toctree-l4"><a class="reference internal" href="tirank.TrainPre.tune_hyperparameters.html">tirank.TrainPre.tune_hyperparameters</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="tirank.Loss.html">tirank.Loss</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#data-handling-preprocessing">Data Handling &amp; Preprocessing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#feature-engineering">Feature Engineering</a></li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#plotting-visualization">Plotting &amp; Visualization</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">TiRank</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../api.html">API Reference</a></li>
      <li class="breadcrumb-item active">tirank.TrainPre</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/generated/tirank.TrainPre.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="tirank-trainpre">
<h1>tirank.TrainPre<a class="headerlink" href="#tirank-trainpre" title="Link to this heading">ÔÉÅ</a></h1>
<dl class="py function">
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">tirank.TrainPre.</span></span><span class="sig-name descname"><span class="pre">Train_one_epoch</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dataloader_A</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dataloader_B</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Cox'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">infer_mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'SC'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adj_A</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adj_B</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pre_patho_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">optimizer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alphas</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[1,</span> <span class="pre">1,</span> <span class="pre">1,</span> <span class="pre">1]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">device</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cpu'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">epoch</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/tirank/TrainPre.html#Train_one_epoch"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Performs a single training epoch for the TiRank multi-task model.</p>
<p>This function iterates through the sc/st data (dataloader_B) and computes
the composite loss against the bulk data (dataloader_A) and regularization
terms (adjacency matrices, MMD, etc.). It then performs a backward
pass and optimizer step.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> (<a class="reference internal" href="tirank.Model.TiRankModel.html#tirank.Model.TiRankModel" title="tirank.Model.TiRankModel"><em>TiRankModel</em></a>) ‚Äì The TiRank model to be trained.</p></li>
<li><p><strong>dataloader_A</strong> (<em>DataLoader</em>) ‚Äì DataLoader for the bulk data (train).</p></li>
<li><p><strong>dataloader_B</strong> (<em>DataLoader</em>) ‚Äì DataLoader for the sc/st data.</p></li>
<li><p><strong>mode</strong> (<em>str</em><em>, </em><em>optional</em>) ‚Äì Analysis mode (‚ÄòCox‚Äô, ‚ÄòClassification‚Äô, ‚ÄòRegression‚Äô).</p></li>
<li><p><strong>infer_mode</strong> (<em>str</em><em>, </em><em>optional</em>) ‚Äì Inference data type (‚ÄòSC‚Äô or ‚ÄòST‚Äô).</p></li>
<li><p><strong>adj_A</strong> (<em>torch.Tensor</em><em>, </em><em>optional</em>) ‚Äì Adjacency matrix for sc/st data
(e.g., from connectivities).</p></li>
<li><p><strong>adj_B</strong> (<em>torch.Tensor</em><em>, </em><em>optional</em>) ‚Äì Adjacency matrix for spatial distance (ST only).</p></li>
<li><p><strong>pre_patho_labels</strong> (<em>pd.Series</em><em>, </em><em>optional</em>) ‚Äì Ground truth pathology labels (ST only).</p></li>
<li><p><strong>optimizer</strong> (<em>torch.optim.Optimizer</em><em>, </em><em>optional</em>) ‚Äì The optimizer.</p></li>
<li><p><strong>alphas</strong> (<em>list</em><em>, </em><em>optional</em>) ‚Äì List of floats to weight the loss components
[reg_loss, bulk_loss, cosine_loss_exp, spatial/patho_loss].</p></li>
<li><p><strong>device</strong> (<em>str</em><em>, </em><em>optional</em>) ‚Äì The compute device (‚Äòcpu‚Äô or ‚Äòcuda‚Äô).</p></li>
<li><p><strong>epoch</strong> (<em>int</em><em>, </em><em>optional</em>) ‚Äì The current epoch number (for prototype loss scheduling).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dictionary of the average loss values for this epoch.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">tirank.TrainPre.</span></span><span class="sig-name descname"><span class="pre">Validate_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dataloader_A</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dataloader_B</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Cox'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">infer_mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'SC'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adj_A</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adj_B</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pre_patho_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alphas</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[1,</span> <span class="pre">1,</span> <span class="pre">1]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">device</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cpu'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/tirank/TrainPre.html#Validate_model"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Performs a single validation step.</p>
<p>This function iterates through the dataloaders, performs a forward pass,
and calculates the total validation loss (without backpropagation).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> (<a class="reference internal" href="tirank.Model.TiRankModel.html#tirank.Model.TiRankModel" title="tirank.Model.TiRankModel"><em>TiRankModel</em></a>) ‚Äì The TiRank model to be evaluated.</p></li>
<li><p><strong>dataloader_A</strong> (<em>DataLoader</em>) ‚Äì DataLoader for the bulk data (validation).</p></li>
<li><p><strong>dataloader_B</strong> (<em>DataLoader</em>) ‚Äì DataLoader for the sc/st data.</p></li>
<li><p><strong>mode</strong> (<em>str</em><em>, </em><em>optional</em>) ‚Äì Analysis mode (‚ÄòCox‚Äô, ‚ÄòClassification‚Äô, ‚ÄòRegression‚Äô).</p></li>
<li><p><strong>infer_mode</strong> (<em>str</em><em>, </em><em>optional</em>) ‚Äì Inference data type (‚ÄòSC‚Äô or ‚ÄòST‚Äô).</p></li>
<li><p><strong>adj_A</strong> (<em>torch.Tensor</em><em>, </em><em>optional</em>) ‚Äì Adjacency matrix for sc/st data.</p></li>
<li><p><strong>adj_B</strong> (<em>torch.Tensor</em><em>, </em><em>optional</em>) ‚Äì Adjacency matrix for spatial distance (ST only).</p></li>
<li><p><strong>pre_patho_labels</strong> (<em>pd.Series</em><em>, </em><em>optional</em>) ‚Äì Ground truth pathology labels (ST only).</p></li>
<li><p><strong>alphas</strong> (<em>list</em><em>, </em><em>optional</em>) ‚Äì List of floats to weight the loss components.</p></li>
<li><p><strong>device</strong> (<em>str</em><em>, </em><em>optional</em>) ‚Äì The compute device (‚Äòcpu‚Äô or ‚Äòcuda‚Äô).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The average validation loss for this epoch.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">tirank.TrainPre.</span></span><span class="sig-name descname"><span class="pre">Reject_With_GMM_Bio</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pred_bulk</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pred_sc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tolerance</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_components</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_components</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/tirank/TrainPre.html#Reject_With_GMM_Bio"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Performs GMM-based rejection for Classification and Cox modes.</p>
<p>This function identifies phenotype-associated clusters by fitting a GMM
to the bulk scores (to find target means 0 and 1) and another GMM
to the sc/st scores, then finding sc/st clusters whose means align
with the bulk targets within a given tolerance.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pred_bulk</strong> (<em>np.ndarray</em>) ‚Äì Predicted scores from the bulk data (n_samples, 1).</p></li>
<li><p><strong>pred_sc</strong> (<em>np.ndarray</em>) ‚Äì Predicted scores from the sc/st data (n_cells, 1).</p></li>
<li><p><strong>tolerance</strong> (<em>float</em>) ‚Äì The maximum distance a sc/st cluster mean can be from
a bulk target mean to be considered aligned.</p></li>
<li><p><strong>min_components</strong> (<em>int</em>) ‚Äì The minimum number of GMM components to try.</p></li>
<li><p><strong>max_components</strong> (<em>int</em>) ‚Äì The maximum number of GMM components to try.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>A binary mask (n_cells, 1) where 1 indicates a cell</dt><dd><p>to be rejected (phenotype-independent) and 0 indicates a cell
to be kept.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">tirank.TrainPre.</span></span><span class="sig-name descname"><span class="pre">Reject_With_GMM_Reg</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pred_bulk</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pred_sc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tolerance</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/tirank/TrainPre.html#Reject_With_GMM_Reg"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Performs GMM-based rejection for Regression mode.</p>
<p>Fits a single-component GMM to both bulk and sc/st scores to find their
means. If the means are too divergent, rejects all cells. Otherwise,
rejects cells that fall outside a tolerance range around the bulk mean.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pred_bulk</strong> (<em>np.ndarray</em>) ‚Äì Predicted scores from the bulk data (n_samples, 1).</p></li>
<li><p><strong>pred_sc</strong> (<em>np.ndarray</em>) ‚Äì Predicted scores from the sc/st data (n_cells, 1).</p></li>
<li><p><strong>tolerance</strong> (<em>float</em>) ‚Äì The tolerance (std dev or max value) to define the
acceptance range around the bulk mean.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A binary mask (n_cells, 1) where 1 indicates rejection.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">tirank.TrainPre.</span></span><span class="sig-name descname"><span class="pre">Reject_With_StrictNumber</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pred_bulk</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pred_sc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tolerance</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/tirank/TrainPre.html#Reject_With_StrictNumber"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Performs rejection based on a strict percentile range.</p>
<p>Fits a 2-component GMM to bulk scores to find means and std deviations.
It then defines an acceptance range based on the percentile (tolerance)
of a normal distribution (e.g., tolerance=0.95 keeps the central 95%
of each bulk cluster).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pred_bulk</strong> (<em>np.ndarray</em>) ‚Äì Predicted scores from the bulk data (n_samples, 1).</p></li>
<li><p><strong>pred_sc</strong> (<em>np.ndarray</em>) ‚Äì Predicted scores from the sc/st data (n_cells, 1).</p></li>
<li><p><strong>tolerance</strong> (<em>float</em>) ‚Äì The percentile of the distribution to keep (e.g., 0.95).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A binary mask (n_cells, 1) where 1 indicates rejection.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">tirank.TrainPre.</span></span><span class="sig-name descname"><span class="pre">objective</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">trial</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_features</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nhead</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nhid1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nhid2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_output</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nlayers</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_pred</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dropout</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_patho</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">encoder_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">train_loader_Bulk</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">val_loader_Bulk</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">train_loader_SC</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adj_A</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adj_B</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pre_patho_labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">device</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">infer_mode</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model_save_path</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/tirank/TrainPre.html#objective"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>The objective function for Optuna hyperparameter optimization.</p>
<p>This function defines the search space for hyperparameters (lr, epochs,
loss weights), builds a model, trains it, and returns the validation loss.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>trial</strong> (<em>optuna.trial.Trial</em>) ‚Äì An Optuna trial object.</p></li>
<li><p><strong>wrapper</strong><strong>)</strong> (<em>(</em><em>All other args are passed from the tune_hyperparameters</em>)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The validation loss for the trial.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">tirank.TrainPre.</span></span><span class="sig-name descname"><span class="pre">tune_hyperparameters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">savePath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">device</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cpu'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_trials</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/tirank/TrainPre.html#tune_hyperparameters"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Runs the Optuna hyperparameter tuning study.</p>
<p>This function loads all necessary data and model parameters, then
initializes and runs an Optuna study to find the best
hyperparameters by minimizing the validation loss.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>savePath</strong> (<em>str</em>) ‚Äì The main project directory path.</p></li>
<li><p><strong>device</strong> (<em>str</em><em>, </em><em>optional</em>) ‚Äì The compute device. Defaults to ‚Äúcpu‚Äù.</p></li>
<li><p><strong>n_trials</strong> (<em>int</em><em>, </em><em>optional</em>) ‚Äì The number of Optuna trials to run.
Defaults to 50.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">tirank.TrainPre.</span></span><span class="sig-name descname"><span class="pre">get_best_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">savePath</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/tirank/TrainPre.html#get_best_model"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Loads the best performing model from the hyperparameter tuning.</p>
<p>This function reads the ‚Äòbest_params.pkl‚Äô file, reconstructs the
corresponding model filename, initializes a new <cite>TiRankModel</cite>
with the saved parameters, and loads the weights.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>savePath</strong> (<em>str</em>) ‚Äì The main project directory path.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The trained TiRank model with the best weights loaded.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="tirank.Model.TiRankModel.html#tirank.Model.TiRankModel" title="tirank.Model.TiRankModel">TiRankModel</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">tirank.TrainPre.</span></span><span class="sig-name descname"><span class="pre">Predict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">savePath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">do_reject</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tolerance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reject_mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'GMM'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/tirank/TrainPre.html#Predict"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Performs inference using the best trained TiRank model.</p>
<p>This function loads the best model, loads the full bulk and sc/st gene
pair matrices, and predicts scores for all samples. It then applies the
chosen rejection (filtering) method to classify cells/spots as
‚ÄòRank+‚Äô, ‚ÄòRank-‚Äô, or ‚ÄòBackground‚Äô. The final results are saved to
‚Äòspot_predict_score.csv‚Äô and ‚Äòfinal_anndata.h5ad‚Äô.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>savePath</strong> (<em>str</em>) ‚Äì The main project directory path.</p></li>
<li><p><strong>mode</strong> (<em>str</em>) ‚Äì The analysis mode (‚ÄòCox‚Äô, ‚ÄòClassification‚Äô, ‚ÄòRegression‚Äô).</p></li>
<li><p><strong>do_reject</strong> (<em>bool</em><em>, </em><em>optional</em>) ‚Äì Whether to perform the GMM-based rejection.
Defaults to True.</p></li>
<li><p><strong>tolerance</strong> (<em>float</em><em>, </em><em>optional</em>) ‚Äì The tolerance parameter for the rejection
method. Defaults to 0.05.</p></li>
<li><p><strong>reject_mode</strong> (<em>str</em><em>, </em><em>optional</em>) ‚Äì The rejection method to use (‚ÄòGMM‚Äô or ‚ÄòStrict‚Äô).
Defaults to ‚ÄúGMM‚Äù.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">tirank.TrainPre.</span></span><span class="sig-name descname"><span class="pre">permute_once</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">Rank_Labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unique_labels</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/tirank/TrainPre.html#permute_once"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Helper function for a single permutation test shuffle.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>Rank_Labels</strong> (<em>list</em>) ‚Äì The list of all ‚ÄòRank_Label‚Äô assignments.</p></li>
<li><p><strong>Labels</strong> (<em>list</em>) ‚Äì The list of all cluster assignments.</p></li>
<li><p><strong>unique_labels</strong> (<em>set</em>) ‚Äì The set of unique cluster labels.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dictionary with permuted counts for each cluster.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">tirank.TrainPre.</span></span><span class="sig-name descname"><span class="pre">AssignPcluster</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_p_values</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/tirank/TrainPre.html#AssignPcluster"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Assigns a final phenotype (‚ÄòRank+‚Äô, ‚ÄòRank-‚Äô, ‚ÄòBackground‚Äô) to clusters.</p>
<p>This assignment is based on the p-values from the permutation test.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>df_p_values</strong> (<em>pd.DataFrame</em>) ‚Äì DataFrame with p-values for ‚ÄòRank+‚Äô,
‚ÄòRank-‚Äô, and ‚ÄòBackground‚Äô enrichment for each cluster.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>A dictionary mapping the final assignment (‚ÄòRank+‚Äô, ‚ÄòRank-‚Äò,</dt><dd><p>‚ÄôBackground‚Äô) to a list of cluster IDs.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">tirank.TrainPre.</span></span><span class="sig-name descname"><span class="pre">Pcluster</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">savePath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clusterColName</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">perm_n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1001</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/tirank/TrainPre.html#Pcluster"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Performs a permutation test to identify significantly enriched clusters.</p>
<p>This function tests whether the observed number of ‚ÄòRank+‚Äô or ‚ÄòRank-‚Äô
labels within any given cluster (e.g., ‚Äòleiden_clusters‚Äô) is
significantly higher than expected by chance.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>savePath</strong> (<em>str</em>) ‚Äì The main project directory path.</p></li>
<li><p><strong>clusterColName</strong> (<em>str</em>) ‚Äì The column name in ‚Äòspot_predict_score.csv‚Äô
that contains the cluster labels (e.g., ‚Äòleiden_clusters‚Äô,
‚Äòpatho_class‚Äô).</p></li>
<li><p><strong>perm_n</strong> (<em>int</em><em>, </em><em>optional</em>) ‚Äì The number of permutations to run.
Defaults to 1001.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py">
<span class="sig-prename descclassname"><span class="pre">tirank.TrainPre.</span></span><span class="sig-name descname"><span class="pre">IdenHub</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">savePath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cateCol1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cateCol2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_spots</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/tirank/TrainPre.html#IdenHub"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Identifies ‚Äúhubs‚Äù by combining two categorical cluster labels.</p>
<p>This function creates a new ‚Äòcombine_cluster‚Äô column by merging two
existing cluster labels (e.g., ‚Äòpatho_class‚Äô and ‚Äòleiden_clusters‚Äô).
It also filters out any combined clusters that have fewer than
<cite>min_spots</cite> members, labeling them as ‚ÄòNA‚Äô.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>savePath</strong> (<em>str</em>) ‚Äì The main project directory path.</p></li>
<li><p><strong>cateCol1</strong> (<em>str</em>) ‚Äì The name of the first categorical column in
‚Äòspot_predict_score.csv‚Äô.</p></li>
<li><p><strong>cateCol2</strong> (<em>str</em>) ‚Äì The name of the second categorical column.</p></li>
<li><p><strong>min_spots</strong> (<em>int</em>) ‚Äì The minimum number of spots required to keep
a combined cluster.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<p class="rubric">Functions</p>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="tirank.TrainPre.AssignPcluster.html#tirank.TrainPre.AssignPcluster" title="tirank.TrainPre.AssignPcluster"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tirank.TrainPre.AssignPcluster</span></code></a></p></td>
<td><p>Assigns a final phenotype ('Rank+', 'Rank-', 'Background') to clusters.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="tirank.TrainPre.IdenHub.html#tirank.TrainPre.IdenHub" title="tirank.TrainPre.IdenHub"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tirank.TrainPre.IdenHub</span></code></a></p></td>
<td><p>Identifies &quot;hubs&quot; by combining two categorical cluster labels.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="tirank.TrainPre.Pcluster.html#tirank.TrainPre.Pcluster" title="tirank.TrainPre.Pcluster"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tirank.TrainPre.Pcluster</span></code></a></p></td>
<td><p>Performs a permutation test to identify significantly enriched clusters.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="tirank.TrainPre.Predict.html#tirank.TrainPre.Predict" title="tirank.TrainPre.Predict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tirank.TrainPre.Predict</span></code></a></p></td>
<td><p>Performs inference using the best trained TiRank model.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="tirank.TrainPre.Reject_With_GMM_Bio.html#tirank.TrainPre.Reject_With_GMM_Bio" title="tirank.TrainPre.Reject_With_GMM_Bio"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tirank.TrainPre.Reject_With_GMM_Bio</span></code></a></p></td>
<td><p>Performs GMM-based rejection for Classification and Cox modes.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="tirank.TrainPre.Reject_With_GMM_Reg.html#tirank.TrainPre.Reject_With_GMM_Reg" title="tirank.TrainPre.Reject_With_GMM_Reg"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tirank.TrainPre.Reject_With_GMM_Reg</span></code></a></p></td>
<td><p>Performs GMM-based rejection for Regression mode.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="tirank.TrainPre.Reject_With_StrictNumber.html#tirank.TrainPre.Reject_With_StrictNumber" title="tirank.TrainPre.Reject_With_StrictNumber"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tirank.TrainPre.Reject_With_StrictNumber</span></code></a></p></td>
<td><p>Performs rejection based on a strict percentile range.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="tirank.TrainPre.Train_one_epoch.html#tirank.TrainPre.Train_one_epoch" title="tirank.TrainPre.Train_one_epoch"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tirank.TrainPre.Train_one_epoch</span></code></a></p></td>
<td><p>Performs a single training epoch for the TiRank multi-task model.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="tirank.TrainPre.Validate_model.html#tirank.TrainPre.Validate_model" title="tirank.TrainPre.Validate_model"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tirank.TrainPre.Validate_model</span></code></a></p></td>
<td><p>Performs a single validation step.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="tirank.TrainPre.get_best_model.html#tirank.TrainPre.get_best_model" title="tirank.TrainPre.get_best_model"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tirank.TrainPre.get_best_model</span></code></a></p></td>
<td><p>Loads the best performing model from the hyperparameter tuning.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="tirank.TrainPre.objective.html#tirank.TrainPre.objective" title="tirank.TrainPre.objective"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tirank.TrainPre.objective</span></code></a></p></td>
<td><p>The objective function for Optuna hyperparameter optimization.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="tirank.TrainPre.permute_once.html#tirank.TrainPre.permute_once" title="tirank.TrainPre.permute_once"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tirank.TrainPre.permute_once</span></code></a></p></td>
<td><p>Helper function for a single permutation test shuffle.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="tirank.TrainPre.tune_hyperparameters.html#tirank.TrainPre.tune_hyperparameters" title="tirank.TrainPre.tune_hyperparameters"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tirank.TrainPre.tune_hyperparameters</span></code></a></p></td>
<td><p>Runs the Optuna hyperparameter tuning study.</p></td>
</tr>
</tbody>
</table>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="tirank.Model.TransformerEncoderModel.html" class="btn btn-neutral float-left" title="tirank.Model.TransformerEncoderModel" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="tirank.TrainPre.AssignPcluster.html" class="btn btn-neutral float-right" title="tirank.TrainPre.AssignPcluster" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, LenisLin.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>